src/SouffleInterface.h

```diff
@@ -376,6 +376,51 @@ class Relation {
      * in the table, set the next element pointer points to the current element itself.
      */
     virtual void purge() = 0;
+
+    /**
+     * Helper function for the wrapper function
+     */
+    template <typename Tuple, size_t N>
+    struct tuple_insert {
+        static void add(const Tuple& t, souffle::tuple& t1) {
+            tuple_insert<Tuple, N - 1>::add(t, t1);
+            t1 << std::get<N - 1>(t);
+        }
+    };
+
+    /**
+     * Helper function for the wrapper function in the first element
+     */
+    template <typename Tuple>
+    struct tuple_insert<Tuple, 1> {
+        static void add(const Tuple& t, souffle::tuple& t1) {
+            t1 << std::get<0>(t);
+        }
+    };
+
+    /**
+     * Insert function with std::tuple as input (wrapper)
+     *
+     * @see Relation::insert()
+     */
+    template <typename... Args>
+    void insert(const std::tuple<Args...>& t) {
+        tuple t1(this);
```

Comment:
What is a `tuple` here? `souffle::tuple` is only forward declared at this point, If you need to use the constructor, can these helpers be moved after the definition of tuple below?

Reply:
The tuple variable here is the variable that going to be inserted into the relation.
Now the wrapper function is in the relation scope, it will be used in client code like `relation->insert(a_tuple)`. If they move after tuple, I think they could only be used like `prog->insert(relation, a_tuple)`, which is the first version of the wrapper.

Reply:
I was asking because the constructor of tuple is called before it is defined. Clang objects to this, since the type is incomplete (and I don't know why gcc is okay with it).
```
In file included from main.cpp:26:
In file included from ./Explain.h:19:
In file included from ./ExplainProvenanceImpl.h:20:
In file included from ./ExplainProvenance.h:21:
./SouffleInterface.h:409:15: error: variable has incomplete type
      'souffle::tuple'
        tuple t1(this);
              ^  
./SouffleInterface.h:34:7: note: forward declaration of 'souffle::tuple'
class tuple;
      ^  
./SouffleInterface.h:421:15: error: variable has incomplete type
      'souffle::tuple'
        tuple t1(this);
              ^  
./SouffleInterface.h:34:7: note: forward declaration of 'souffle::tuple'
class tuple;
      ^  
```

Reply:
Oh... I get what you mean... The code work for me as well, I didn't notice that...
If I move the wrapper below tuple class, I couldn't think a way that the wrapper could call as `relation->insert(a_tuple)`, I tried to modify CompiledSouffle.h but soon I realised there is no virtual-template method in C++, std::tuple iteration relied on template, I could not leave a virtual method in Realtion class, do you have any suggestion?

Reply:
Currently we forward declare tuple, then define relation, then define tuple. Would it work if we forward declared relation, then define tuple, then define relation?

Reply:
It won't work, since something in "tuple" uses things from "relation".

Reply:
Ok, so, quick fix. Move the wrapper functions to `SouffleProgram`, and pass them the relation as an extra parameter.